server/mcp/user_client.go

```diff
@@ -0,0 +1,284 @@
+// Copyright (c) 2023-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package mcp
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"maps"
+	"time"
+
+	"github.com/invopop/jsonschema"
+	"github.com/mark3labs/mcp-go/client"
+	"github.com/mark3labs/mcp-go/mcp"
+	"github.com/mattermost/mattermost-plugin-ai/server/llm"
+	"github.com/mattermost/mattermost/server/public/pluginapi"
+	orderedmap "github.com/wk8/go-ordered-map/v2"
+)
+
+const MMUserIDHeader = "X-Mattermost-UserID"
+
+// ServerConnection represents the connection to a single MCP server
+type ServerConnection struct {
+	client   *client.SSEMCPClient
+	serverID string
+	tools    map[string]mcp.Tool
+}
+
+// ServerConfig contains the configuration for a single MCP server
+type ServerConfig struct {
+	BaseURL string            `json:"baseURL"`
+	Headers map[string]string `json:"headers,omitempty"`
+}
+
+// ToolDefinition represents a tool provided by an MCP server
+type ToolDefinition struct {
+	tool     mcp.Tool
+	serverID string
+}
+
+// UserClient represents a per-user MCP client with multiple server connections
+type UserClient struct {
+	clients      map[string]*ServerConnection
+	toolDefs     map[string]ToolDefinition
+	lastActivity time.Time
+	userID       string
+	log          pluginapi.LogService
+}
+
+// ConnectToAllServers initializes connections to all provided servers
+func (c *UserClient) ConnectToAllServers(servers map[string]ServerConfig) error {
+	// Initialize clients for each server
+	for serverID, serverConfig := range servers {
+		if serverConfig.BaseURL == "" {
+			c.log.Warn("Skipping MCP server with empty BaseURL", "serverID", serverID)
+			continue
+		}
+
+		if err := c.connectToServer(context.Background(), serverID, serverConfig); err != nil {
+			c.log.Error("Failed to connect to MCP server", "userID", c.userID, "serverID", serverID, "error", err)
+			continue
+		}
+	}
+
+	// If no servers were successfully connected, return error
+	if len(c.clients) == 0 {
```

Comment:
There is a case where you run "ConnectToAllServers" and the servers list is empty that this could happen, and I don't think it makes sense. The other thing is, this function doesn't error even if most of the connectToServer failed. What I understand is more forgiving, and still it is shown in the logs, but I'm not sure about that behavior. I think I like the, "best-effort" approach that you are following here, but I want to raise the concern about it.

Reply:
Will add a check for 0 and not error out. 
I don't really see a better approach here. I think the way to resolve this is a UI for admins to see the status of MCP servers.